models/inventory/helpers.ts

```diff
@@ -49,3 +49,211 @@ async function validateItemRowBatch(
+    ModelNameEnum.Item,
+    item,
+    'hasSerialNo'
+  );
+
+  if (hasSerialNo && !doc.serialNo) {
+    throw new ValidationError(
+      [
+        doc.fyo.t`Serial No not set for row ${idx}.`,
+        doc.fyo.t`Serial No is enabled for Item ${item}`,
+      ].join(' ')
+    );
+  }
+
+  if (!hasSerialNo && doc.serialNo) {
+    throw new ValidationError(
+      [
+        doc.fyo.t`Serial No set for row ${idx}.`,
+        doc.fyo.t`Serial No is not enabled for Item ${item}`,
+      ].join(' ')
+    );
+  }
+
+  if (!hasSerialNo) return;
+
+  const serialNos = getSerialNumbers(doc.serialNo as string);
+
+  if (serialNos.length !== doc.quantity) {
+    throw new ValidationError(
+      t`${doc.quantity!} Serial Numbers required for ${doc.item!}. You have provided ${
+        serialNos.length
+      }.`
+    );
+  }
+
+  for (const serialNo of serialNos) {
+    const { name, status, item } = await doc.fyo.db.get(
+      ModelNameEnum.SerialNo,
+      serialNo,
+      ['name', 'status', 'item']
+    );
+
+    if (movementType == 'MaterialIssue') {
+      await validateSNMaterialIssue(
+        doc,
+        name as string,
+        item as string,
+        serialNo,
+        status as string
+      );
+    }
+
+    if (movementType == 'MaterialReceipt') {
+      await validateSNMaterialReceipt(
+        doc,
+        name as string,
+        serialNo,
+        status as string
+      );
+    }
+
+    if (movementType === 'Shipment') {
+      await validateSNShipment(doc, serialNo);
+    }
+
+    if (doc.parentSchemaName === 'PurchaseReceipt') {
+      await validateSNPurchaseReceipt(
+        doc,
+        name as string,
+        serialNo,
+        status as string
+      );
+    }
+  }
+}
+
+export const getSerialNumbers = (serialNo: string): string[] => {
+  return serialNo ? serialNo.split('\n') : [];
+};
+
+export const updateSerialNoStatus = async (
+  doc: Doc,
+  items: StockMovementItem[] | ShipmentItem[],
+  newStatus: string
+) => {
+  for (const item of items) {
+    const serialNos = getSerialNumbers(item.serialNo!);
+    if (!serialNos.length) break;
+
+    for (const serialNo of serialNos) {
+      await doc.fyo.db.update(ModelNameEnum.SerialNo, {
+        name: serialNo,
+        status: newStatus,
+      });
+    }
+  }
+};
+
+
+const validateSNMaterialReceipt = async (
+  doc: Doc,
+  name: string,
+  serialNo: string,
+  status: string
+) => {
+  if (name === undefined) {
+    const values = {
+      name: serialNo,
+      item: doc.item,
+      party: doc.parentdoc?.party as string,
+    };
+    (
+      await doc.fyo.doc
+        .getNewDoc(ModelNameEnum.SerialNo, values as DocValueMap)
+        .sync()
+    ).submit();
+  }
+
+  if (status && status !== 'Inactive') {
+    throw new ValidationError(t`SerialNo ${serialNo} status is not Inactive`);
+  }
+};
+
+const validateSNPurchaseReceipt = async (
+  doc: Doc,
+  name: string,
+  serialNo: string,
+  status: string
+) => {
+  if (name === undefined) {
```

Comment:
Why are you even checking if name will be undefined here? The calling function already has a type restraint that `name` has to be a string!

This is bad!